src/coreclr/jit/gentree.cpp

```diff
@@ -30512,6 +30486,73 @@ GenTree* Compiler::gtFoldExprHWIntrinsic(GenTreeHWIntrinsic* tree)
         }
     }
 
+    if (op3 != nullptr)
+    {
+        switch (ni)
+        {
+            case NI_Vector128_ConditionalSelect:
+#if defined(TARGET_XARCH)
+            case NI_Vector256_ConditionalSelect:
+            case NI_Vector512_ConditionalSelect:
+#elif defined(TARGET_ARM64)
+            case NI_Vector64_ConditionalSelect:
+            case NI_Sve_ConditionalSelect:
+#endif
+            {
+                if (cnsNode != op1)
+                {
+                    break;
+                }
+
+                if (op1->IsVectorAllBitsSet() || op1->IsMaskAllBitsSet())
+                {
+                    if ((op3->gtFlags & GTF_SIDE_EFFECT) == 0)
+                    {
+                        // op3 has no side effects, so we can return op2 directly
+                        return op2;
+                    }
+
+                    if ((op2->gtFlags & GTF_SIDE_EFFECT) == 0)
+                    {
+                        // op2 has no side effects, but op3 does, we need to wrap
+                        // so that the side effects for op3 are preserved
+                        return gtWrapWithSideEffects(op2, op3, GTF_ALL_EFFECT);
+                    }
```

Comment:
can you explain why we need to do this?

Reply:
If we didn't we'd end up losing track of `op3` and its side effects would no longer evaluate, since it would no longer be the child of one of the statement nodes in the basic block

> An HIR block is composed of a doubly-linked list of statement nodes (Statement), each of which references a single expression tree (m_rootNode). The GenTree nodes in this tree execute in "tree order", which is defined as the order produced by a depth-first, left-to-right traversal of the tree, with one notable exception

We don't want to insert new statements arbitrarily and `gtFoldExpr` is called from many different phases, so it tends to do the very simple things and otherwise leaves the rest to VN, CSE, and AssertionProp

Reply:
Just knowing that `op2` is side effect free here is not sufficient to allow reordering it and `op3`; the side effects of `op3` may still affect the value of `op2`. For example, `op2` could be `LCL_VAR(V00)` and `op3` could be something that stores to `V00`.

Reply:
Ah, yeah, that's right. Not sure why I didn't remember that scenario. Fixed.